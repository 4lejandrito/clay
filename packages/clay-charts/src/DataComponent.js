import {isFunction, isObject, isString} from 'metal';
import {Config} from 'metal-state';
import types from './utils/types';

/**
 * DataComponent prototype.
 *
 * @mixin
 */
const DataComponent = {
	/**
	 * Check's the data or columns option and resolves this `data_` accordingly.
	 * @return {Promise}
	 * @param {?Array|Function|String} data the data to resolve.
	 * @protected
	 */
	_resolveData(data) {
		if (Array.isArray(data) || (isObject(data) && !isFunction(data))) {
			return Promise.resolve(data);
		} else if (isFunction(data)) {
			return data().then(val => val);
		} else if (isString(data)) {
			return fetch(data, {cors: 'cors'})
				.then(res => res.json())
				.then(res => res.data);
		} else {
			Promise.reject(`Invalid type for data: ${data}`);
		}
	},

	/**
	 * Sets up the polling interval.
	 */
	_setupPolling() {
		if (this.pollingInterval) {
			if (this._pollingInterval) {
				clearInterval(this._pollingInterval);
			}

			this._pollingInterval = setInterval(() => {
				this._updateData(this.data);
			}, this.pollingInterval);
		}
	},
};

/**
 * State definition.
 * @static
 * @type {!Object}
 */
DataComponent.STATE = {
	/**
	 * Data that will be rendered to the chart.
	 * @instance
	 * @memberof ChartBase
	 * @type {?Array|undefined}
	 * @default undefined
	 */
	data: Config.oneOfType([
		Config.arrayOf(
			Config.shapeOf({
				axis: Config.oneOf(['y', 'y2']),
				class: Config.string(),
				color: Config.string(),
				data: Config.array().required(),
				hide: Config.bool(),
				id: Config.required().string(),
				name: Config.string(),
				regions: Config.array(),
				type: Config.oneOf(types.all),
				x: Config.string(),
			})
		),
		Config.object(),
		Config.func(),
		Config.string(),
	]),

	/**
	 * Set an interval (in ms) to fetch the data.
	 * @instance
	 * @memberof ChartBase
	 * @type {?Number}
	 * @default undefined
	 */
	pollingInterval: Config.number(),
};

export {DataComponent};
export default DataComponent;
