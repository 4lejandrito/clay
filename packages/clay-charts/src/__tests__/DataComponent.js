const mockData = {data: {}};

import DataComponent from '../DataComponent';

describe('DataComponent', () => {
	beforeEach(() => {
		global.fetch = jest.fn().mockImplementation(() =>
			Promise.resolve({
				json: jest.fn().mockImplementation(() => Promise.resolve(mockData)),
			})
		);
	});

	afterEach(() => {
		global.fetch.mockRestore();
	});

	it('should accept data as a string', () => {
		const comp = new DataComponent({
			data: './data/world-low-res.geo.json',
		});

		expect(comp).toMatchSnapshot();
	});

	it('should accept data as an object', () => {
		const component = new DataComponent({
			data: mockData.data,
		});

		expect(component).toMatchSnapshot();
	});

	it('should accept data as a function', () => {
		const component = new DataComponent({
			data: () => {
				return new Promise(resolve => {
					setTimeout(() => {
						resolve(mockData.data);
					}, 500);
				});
			},
		});
		expect(component).toMatchSnapshot();
	});

	it('should warn if data type is invalid', () => {
		global.console.error = jest.fn();

		const createComponent = () => {
			return new DataComponent({
				data: 123,
			});
		};

		createComponent();
		expect(global.console.error.mock.calls[0][0]).toBeDefined();
		global.console.error.mockRestore();
	});

	it('should support `pollingInterval`', done => {
		const component = new DataComponent({
			data: mockData.data,
			pollingInterval: 100,
		});

		const updateData = (DataComponent.prototype._updateData = jest.fn());

		component._setupPolling();
		component._setupPolling();

		expect(component._pollingInterval).toBeDefined();

		setTimeout(() => {
			expect(updateData.mock.calls[0]).toBeDefined();
			updateData.mockRestore();
			done();
		}, 500);
	});

	it('should clean up `pollingInterval`', () => {
		const component = new DataComponent({
			data: mockData.data,
			pollingInterval: 100,
		});
		component._setupPolling();
		expect(component._pollingInterval).toBeDefined();
		component.cleanup();
		expect(component._pollingInterval).toBe(null);
	});
});
