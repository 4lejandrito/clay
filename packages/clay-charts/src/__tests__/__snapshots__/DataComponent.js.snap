// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataComponent should accept data as a function 1`] = `<div></div>`;

exports[`DataComponent should accept data as a string 1`] = `<div></div>`;

exports[`DataComponent should accept data as an object 1`] = `<div></div>`;
